src/services/azureNamingService.ts

```diff
@@ -0,0 +1,228 @@
+import md5 from "md5";
+import Serverless from "serverless";
+import configConstants from "../config";
+import { ArmResourceType } from "../models/armTemplates";
+import { DeploymentConfig, ServerlessAzureConfig, ServerlessAzureOptions } from "../models/serverless";
+import { Utils } from "../shared/utils";
+
+export class AzureNamingService {
+
+  private config: ServerlessAzureConfig;
+
+  public constructor(
+    private serverless: Serverless,
+    private options: ServerlessAzureOptions = { stage: null, region: null }
+  ) {
+    this.setDefaultValues();
+    this.config = serverless.service as any;
+  }
+
+  /**
+   * Name of Function App Service
+   */
+  public getServiceName(): string {
+    return this.config.service;
+  }
+
+  /**
+   * Name of Azure Region for deployment
+   */
+  public getRegion(): string {
+    return this.options.region || this.config.provider.region;
+  }
+
+  /**
+   * Name of current deployment stage
+   */
+  public getStage(): string {
+    return this.options.stage || this.config.provider.stage;
+  }
+
+  /**
+   * Name of prefix for service
+   */
+  public getPrefix(): string {
+    return this.config.provider.prefix;
+  }
+
+  /**
+   * Name of current resource group
+   */
+  public getResourceGroupName(): string {
+    const regionName = Utils.createShortAzureRegionName(this.getRegion());
+    const stageName = Utils.createShortStageName(this.getStage());
+
+    return this.options.resourceGroup
+      || this.config.provider.resourceGroup
+      || `${this.getPrefix()}-${regionName}-${stageName}-${this.getServiceName()}-rg`;
+  }
+
+  /**
+   * Name of current ARM deployment
+   */
+  public getDeploymentName(): string {
+    const name = this.config.provider.deploymentName || `${this.getResourceGroupName()}-deployment`;
+    return this.rollbackConfiguredName(name);
+  }
+
+  /**
+   * Name of artifact uploaded to blob storage
+   */
+  public getArtifactName(deploymentName: string): string {
+    return `${deploymentName
+      .replace("rg-deployment", "artifact")
+      .replace("deployment", "artifact")}.zip`;
+  }
+
+  /**
+   * Get name of Azure resource
+   * @param resource ARM Resource to name
+   */
+  public getResourceName(resource: ArmResourceType): string {
```

Comment:
I'm also uneasy about this change.  Currently the "naming convention" is controlled by the owning resource and the business logic / naming convention is close to the resource that solves other concerns about that resource.  Moving the resource specific implementation out of the resource into another service with an every growing number of functions seems off to me.  Let's discuss the overall direction here.